From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chosen_1st <haizhongdelan@outlook.com>
Date: Sat, 12 Aug 2023 20:46:35 +0800
Subject: [PATCH] Add Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 9af4bc16da09e59009c47911219e99450cdf2aa5..ab04c2c720989cc97dd3d3617e1bc3821b34c038 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2196,6 +2196,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         {
             throw new UnsupportedOperationException("Not supported yet.");
         }
+        // Augustus start - Add Config and Command
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getAugustusConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Augustus end - Add Config and Command
         // Paper end
 
         /**
