From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chosen <haizhongdelan@outlook.com>
Date: Sat, 12 Aug 2023 20:46:35 +0800
Subject: [PATCH] Add Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 68206cf0178c26c0f528a1e14a5fb4e9ad410369..74f9d24c93d34047f06b7600c961d98cd5b89883 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2079,6 +2079,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         {
             throw new UnsupportedOperationException("Not supported yet.");
         }
+        // Augustus start - Add Config and Command
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getAugustusConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Augustus end - Add Config and Command
         // Paper end
 
         /**
