From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chosen_1st <haizhongdelan@outlook.com>
Date: Sat, 12 Aug 2023 20:46:35 +0800
Subject: [PATCH] Add Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d0c634629aa0b6bac0da93655dd86ad3aea0ce30..58699c01db66b3972ccd6b9bef30901d6a0977bd 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2188,6 +2188,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         {
             throw new UnsupportedOperationException("Not supported yet.");
         }
+        // Augustus start - Add Config and Command
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getAugustusConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Augustus end - Add Config and Command
         // Paper end
 
         /**
