From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chosen_1st <haizhongdelan@outlook.com>
Date: Sat, 12 Aug 2023 20:46:35 +0800
Subject: [PATCH] Add Config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 8d8fe04e6b09d2a5b1cc05002073df5c58cdcb96..a2ad9f6b31ae797283031ea7dbd8cbf6c5bb5d2f 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2119,6 +2119,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         {
             throw new UnsupportedOperationException("Not supported yet.");
         }
+        // Augustus start - Add Config and Command
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getAugustusConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Augustus end - Add Config and Command
         // Paper end
 
         /**
