From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chosen_1st <haizhongdelan@outlook.com>
Date: Sat, 27 Jan 2024 20:16:38 +0800
Subject: [PATCH] Toggle safe teleportation


diff --git a/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java b/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
index d7d35d41dabc1089b14dd16b794e056d80bfc794..b318bb28ea34c0d63e93b8c65798ca50551dbc39 100644
--- a/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
+++ b/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
@@ -120,4 +120,9 @@ public class AugustusWorldConfig {
         final Map<String, Object> value = getMap("world-settings." + worldName + "." + path, null);
         return value.isEmpty() ? fallback : value;
     }
+    public boolean safeteleporting = true;
+
+    private void Settings() {
+        safeteleporting = getBoolean("settings.safe-teleportation", safeteleporting);
+    }
 }
diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 2c06f3ebf7e1069727387bfc60db30c958c14b5a..ce64138e5328f9082e931caba0a7abf3e358d6e8 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -443,6 +443,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
     public BlockPos portalBlock;
     public ServerLevel portalWorld;
     public void tickEndPortal() {
+        // Augustus start - Toggle safe teleportation
+        if (!this.level.augustusConfig.safeteleporting) {
+            return;
+        }
+        // Augustus end - Toggle safe teleportation
         BlockPos pos = this.portalBlock;
         ServerLevel world = this.portalWorld;
         this.portalBlock = null;
@@ -3172,7 +3177,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, S
             }
 
             this.processPortalCooldown();
-            this.tickEndPortal(); // Paper - make end portalling safe
+            // Augustus start - Toggle safe teleportation
+            if (this.level.augustusConfig.safeteleporting) {
+                this.tickEndPortal(); // Paper - make end portalling safe
+            }
+            // Augustus end - Toggle safe teleportation
         }
     }
 
diff --git a/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java b/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
index a0c1db8cfebaa0344012cc0af18d6231cdcdcbb8..fddf94bc6a38ed1823401f04bb8a31a541bd2619 100644
--- a/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/EndPortalBlock.java
@@ -61,10 +61,23 @@ public class EndPortalBlock extends BaseEntityBlock {
                 // return; // CraftBukkit - always fire event in case plugins wish to change it
             }
 
-            // Paper start - move all of this logic into portal tick
-            entity.portalWorld = ((ServerLevel)world);
-            entity.portalBlock = pos.immutable();
-            // Paper end - move all of this logic into portal tick
+            // Augustus start - Toggle safe teleportation
+            if (entity.level().augustusConfig.safeteleporting) {
+                // Paper start - move all of this logic into portal tick
+                entity.portalWorld = ((ServerLevel) world);
+                entity.portalBlock = pos.immutable();
+                // Paper end - move all of this logic into portal tick
+            } else if (!entity.level().augustusConfig.safeteleporting) {
+                EntityPortalEnterEvent event = new EntityPortalEnterEvent(entity.getBukkitEntity(), new org.bukkit.Location(world.getWorld(), pos.getX(), pos.getY(), pos.getZ()));
+                world.getCraftServer().getPluginManager().callEvent(event);
+
+                if (entity instanceof ServerPlayer) {
+                    ((ServerPlayer) entity).changeDimension(worldserver, PlayerTeleportEvent.TeleportCause.END_PORTAL);
+                    return;
+                }
+                entity.changeDimension(worldserver);
+            }
+            // Augustus end - Toggle safe teleportation
         }
 
     }
