From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chosen_1st <haizhongdelan@outlook.com>
Date: Sat, 27 Jan 2024 20:17:51 +0800
Subject: [PATCH] Toggle sand duping


diff --git a/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java b/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
index b318bb28ea34c0d63e93b8c65798ca50551dbc39..b83024f4a8b7fd2e2e651624d2568337fa77ae9d 100644
--- a/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
+++ b/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
@@ -121,8 +121,10 @@ public class AugustusWorldConfig {
         return value.isEmpty() ? fallback : value;
     }
     public boolean safeteleporting = true;
+    public boolean sandduping = false;
 
     private void Settings() {
         safeteleporting = getBoolean("settings.safe-teleportation", safeteleporting);
+        sandduping = getBoolean("settings.sand-duping", sandduping);
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index 9daf8aa557d9f4fdbcc138a47892ea5a061dd877..3dca20c5ff81ddc09a2258290e6d42ed3a968c4e 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -133,7 +133,7 @@ public class FallingBlockEntity extends Entity {
     @Override
     public void tick() {
         // Paper start - fix sand duping
-        if (this.isRemoved()) {
+        if (!this.level().augustusConfig.sandduping && this.isRemoved()) { // Augustus - Toggle sand duping
             return;
         }
         // Paper end - fix sand duping
@@ -149,7 +149,7 @@ public class FallingBlockEntity extends Entity {
 
             this.move(MoverType.SELF, this.getDeltaMovement());
             // Paper start - fix sand duping
-            if (this.isRemoved()) {
+            if (!this.level().augustusConfig.sandduping && this.isRemoved()) { // Augustus - Toggle sand duping
                 return;
             }
             // Paper end - fix sand duping
