From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chosen_1st <haizhongdelan@outlook.com>
Date: Sat, 18 Nov 2023 19:39:15 +0800
Subject: [PATCH] Toggle sand duping


diff --git a/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java b/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
index d06608dbde346b4f24a0ad7e077e740f825c0f24..a6bf5dedf405cbcfa2391eb8eefe00a4d39431bc 100644
--- a/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
+++ b/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
@@ -122,8 +122,10 @@ public class AugustusWorldConfig {
     }
 
     public boolean safeteleporting = true;
+    public boolean sandduping = false;
 
     private void Settings() {
         safeteleporting = getBoolean("settings.safe-teleportation", safeteleporting);
+        sandduping = getBoolean("settings.sand-duping", sandduping);
     }
 }
diff --git a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
index e6f75a9cac46c8e3ddba664a9d5b27b665a94cb4..33a2d2839b3bbee00a4858148102417d6aaea733 100644
--- a/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -133,7 +133,7 @@ public class FallingBlockEntity extends Entity {
     @Override
     public void tick() {
         // Paper start - fix sand duping
-        if (this.isRemoved()) {
+        if (!this.level().augustusConfig.sandduping && this.isRemoved()) { // Augustus - Toggle sand duping
             return;
         }
         // Paper end - fix sand duping
@@ -150,7 +150,7 @@ public class FallingBlockEntity extends Entity {
             this.move(MoverType.SELF, this.getDeltaMovement());
 
             // Paper start - fix sand duping
-            if (this.isRemoved()) {
+            if (!this.level().augustusConfig.sandduping && this.isRemoved()) { // Augustus - Toggle sand duping
                 return;
             }
             // Paper end - fix sand duping
