From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chosen_1st <haizhongdelan@outlook.com>
Date: Sat, 27 Jan 2024 20:19:54 +0800
Subject: [PATCH] Toggle tripwire update


diff --git a/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java b/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
index b83024f4a8b7fd2e2e651624d2568337fa77ae9d..2c9ac62ca286c6c81eb4b4f7ade1b8b5694f6cb5 100644
--- a/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
+++ b/src/main/java/cc/skydom/server/chosen/AugustusWorldConfig.java
@@ -122,9 +122,11 @@ public class AugustusWorldConfig {
     }
     public boolean safeteleporting = true;
     public boolean sandduping = false;
+    public boolean tripwireupdate = false;
 
     private void Settings() {
         safeteleporting = getBoolean("settings.safe-teleportation", safeteleporting);
         sandduping = getBoolean("settings.sand-duping", sandduping);
+        tripwireupdate = getBoolean("settings.tripwire-update", tripwireupdate);
     }
 }
diff --git a/src/main/java/net/minecraft/world/level/block/TripWireHookBlock.java b/src/main/java/net/minecraft/world/level/block/TripWireHookBlock.java
index 7f2dcf6a9e69779e6f898284b58fb1e32902000c..0af5336a2534874244afcca9373e3b803e26754a 100644
--- a/src/main/java/net/minecraft/world/level/block/TripWireHookBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/TripWireHookBlock.java
@@ -162,8 +162,15 @@ public class TripWireHookBlock extends Block {
                     boolean flag7 = (Boolean) iblockdata2.getValue(TripWireBlock.POWERED);
 
                     flag5 |= flag6 && flag7;
-                    if (k != i || !tripWireBeingRemoved || !flag6) // Paper - fix tripwire state inconsistency; don't update the tripwire again if being removed and not disarmed
-                    aiblockdata[k] = iblockdata2;
+                    // Augustus start - Toggle tripwire update
+                    if (world.augustusConfig.tripwireupdate) {
+                        aiblockdata[k] = iblockdata2; // Augustus - always update it
+                    } else {
+                        if (k != i || !tripWireBeingRemoved || !flag6) { // Paper - don't update the tripwire again if being removed and not disarmed {
+                            aiblockdata[k] = iblockdata2;
+                        }
+                    }
+                    // Augustus end - Toggle tripwire update
                     if (k == i) {
                         world.scheduleTick(pos, block, 10);
                         flag4 &= flag6;
