From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Chosen_1st <haizhongdelan@outlook.com>
Date: Mon, 14 Aug 2023 00:27:48 +0800
Subject: [PATCH] Toggle Protect Bedrock and End Portal Frames from being
 destroyed


diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
index 8f97c9df726ac20cfce7bdddd5dd4f8c5aa76c35..416cfb0b14172b03e77d2861ef1077b0f64dd343 100644
--- a/src/main/java/net/minecraft/world/level/Explosion.java
+++ b/src/main/java/net/minecraft/world/level/Explosion.java
@@ -168,7 +168,11 @@ public class Explosion {
                         for (float f1 = 0.3F; f > 0.0F; f -= 0.22500001F) {
                             BlockPos blockposition = BlockPos.containing(d4, d5, d6);
                             BlockState iblockdata = this.level.getBlockState(blockposition);
-                            if (!iblockdata.isDestroyable()) continue; // Paper
+                            // Augustus start - Toggle Protect Bedrock and End Portal Frames from being destroyed
+                            if (this.level.augustusConfig.preventdestroyingunbreakableblocks) {
+                                if (!iblockdata.isDestroyable()) continue; // Paper
+                            }
+                            // Augustus end - Toggle Protect Bedrock and End Portal Frames from being destroyed
                             FluidState fluid = iblockdata.getFluidState(); // Paper
 
                             if (!this.level.isInWorldBounds(blockposition)) {
@@ -379,7 +383,8 @@ public class Explosion {
                 }
                 // CraftBukkit end
 
-                if (!iblockdata.isAir() && iblockdata.isDestroyable()) { // Paper
+                // Augustus start - Toggle Protect Bedrock and End Portal Frames from being destroyed
+                if (!this.level.augustusConfig.preventdestroyingunbreakableblocks && !iblockdata.isAir()) {
                     BlockPos blockposition1 = blockposition.immutable();
 
                     this.level.getProfiler().push("explosion_blocks");
@@ -405,7 +410,36 @@ public class Explosion {
                     this.level.setBlock(blockposition, Blocks.AIR.defaultBlockState(), 3);
                     block.wasExploded(this.level, blockposition, this);
                     this.level.getProfiler().pop();
+                } else {
+                    if (!iblockdata.isAir() && iblockdata.isDestroyable()) { // Paper
+                        BlockPos blockposition1 = blockposition.immutable();
+
+                        this.level.getProfiler().push("explosion_blocks");
+                        if (block.dropFromExplosion(this)) {
+                            Level world = this.level;
+
+                            if (world instanceof ServerLevel) {
+                                ServerLevel worldserver = (ServerLevel) world;
+                                BlockEntity tileentity = iblockdata.hasBlockEntity() ? this.level.getBlockEntity(blockposition) : null;
+                                LootParams.Builder lootparams_a = (new LootParams.Builder(worldserver)).withParameter(LootContextParams.ORIGIN, Vec3.atCenterOf(blockposition)).withParameter(LootContextParams.TOOL, ItemStack.EMPTY).withOptionalParameter(LootContextParams.BLOCK_ENTITY, tileentity).withOptionalParameter(LootContextParams.THIS_ENTITY, this.source);
+
+                                if (yield < 1.0F) { // CraftBukkit - add yield
+                                    lootparams_a.withParameter(LootContextParams.EXPLOSION_RADIUS, 1.0F / yield); // CraftBukkit - add yield
+                                }
+
+                                iblockdata.spawnAfterBreak(worldserver, blockposition, ItemStack.EMPTY, flag2);
+                                iblockdata.getDrops(lootparams_a).forEach((itemstack) -> {
+                                    Explosion.addBlockDrops(objectarraylist, itemstack, blockposition1);
+                                });
+                            }
+                        }
+
+                        this.level.setBlock(blockposition, Blocks.AIR.defaultBlockState(), 3);
+                        block.wasExploded(this.level, blockposition, this);
+                        this.level.getProfiler().pop();
+                    }
                 }
+                // Augustus end - Toggle Protect Bedrock and End Portal Frames from being destroyed
             }
 
             objectlistiterator = objectarraylist.iterator();
diff --git a/src/main/java/net/minecraft/world/level/Level.java b/src/main/java/net/minecraft/world/level/Level.java
index b1c81c38a43c0e5edc1026e9545fcadd446d0e02..91a5d5fac7376f617e95ab64a985b217c7b61e31 100644
--- a/src/main/java/net/minecraft/world/level/Level.java
+++ b/src/main/java/net/minecraft/world/level/Level.java
@@ -608,10 +608,14 @@ public abstract class Level implements LevelAccessor, AutoCloseable {
     public boolean setBlock(BlockPos pos, BlockState state, int flags, int maxUpdateDepth) {
         // CraftBukkit start - tree generation
         if (this.captureTreeGeneration) {
-            // Paper start
-            BlockState type = getBlockState(pos);
-            if (!type.isDestroyable()) return false;
-            // Paper end
+            // Augustus start - Toggle Protect Bedrock and End Portal Frames from being destroyed
+            if (this.augustusConfig.preventdestroyingunbreakableblocks) {
+                // Paper start
+                BlockState type = getBlockState(pos);
+                if (!type.isDestroyable()) return false;
+                // Paper end
+            }
+            // Augustus end - Toggle Protect Bedrock and End Portal Frames from being destroyed
             CraftBlockState blockstate = this.capturedBlockStates.get(pos);
             if (blockstate == null) {
                 blockstate = CapturedBlockState.getTreeBlockState(this, pos, flags);
diff --git a/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index 58d1f76544d7a8d196f75438579266664e5d7404..3930e3ef543553519be5da5acdd0168dc99f1b68 100644
--- a/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -199,12 +199,16 @@ public class PistonBaseBlock extends DirectionalBlock {
     @Override
     public boolean triggerEvent(BlockState state, Level world, BlockPos pos, int type, int data) {
         Direction enumdirection = (Direction) state.getValue(PistonBaseBlock.FACING);
-        // Paper start - prevent retracting when we're facing the wrong way (we were replaced before retraction could occur)
-        Direction directionQueuedAs = Direction.from3DDataValue(data & 7); // Paper - copied from below
-        if (!io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPermanentBlockBreakExploits && enumdirection != directionQueuedAs) {
-            return false;
+        // Augustus start - Toggle Protect Bedrock and End Portal Frames from being destroyed
+        if (world.augustusConfig.preventdestroyingunbreakableblocks) {
+            // Paper start - prevent retracting when we're facing the wrong way (we were replaced before retraction could occur)
+            Direction directionQueuedAs = Direction.from3DDataValue(data & 7); // Paper - copied from below
+            if (!io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPermanentBlockBreakExploits && enumdirection != directionQueuedAs) {
+                return false;
+            }
+            // Paper end - prevent retracting when we're facing the wrong way
         }
-        // Paper end - prevent retracting when we're facing the wrong way
+        // Augustus end - Toggle Protect Bedrock and End Portal Frames from being destroyed
         BlockState iblockdata1 = (BlockState) state.setValue(PistonBaseBlock.EXTENDED, true);
 
         if (!world.isClientSide) {
@@ -245,7 +249,13 @@ public class PistonBaseBlock extends DirectionalBlock {
             }
             // Paper end
             world.setBlock(pos, iblockdata2, 20);
-            world.setBlockEntity(MovingPistonBlock.newMovingBlockEntity(pos, iblockdata2, (BlockState) this.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.from3DDataValue(data & 7)), enumdirection, false, true)); // Paper - diff on change
+            // Augustus start - Toggle Protect Bedrock and End Portal Frames from being destroyed
+            if (!world.augustusConfig.preventdestroyingunbreakableblocks) {
+                world.setBlockEntity(MovingPistonBlock.newMovingBlockEntity(pos, iblockdata2, (BlockState) this.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.from3DDataValue(data & 7)), enumdirection, false, true));
+            } else {
+                world.setBlockEntity(MovingPistonBlock.newMovingBlockEntity(pos, iblockdata2, (BlockState) this.defaultBlockState().setValue(PistonBaseBlock.FACING, Direction.from3DDataValue(data & 7)), enumdirection, false, true)); // Paper - diff on change
+            }
+            // Augustus end - Toggle Protect Bedrock and End Portal Frames from being destroyed
             world.blockUpdated(pos, iblockdata2.getBlock());
             iblockdata2.updateNeighbourShapes(world, pos, 2);
             if (this.isSticky) {
@@ -281,14 +291,20 @@ public class PistonBaseBlock extends DirectionalBlock {
                     }
                 }
             } else {
-                // Paper start - fix headless pistons breaking blocks
-                BlockPos headPos = pos.relative(enumdirection);
-                if (io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPermanentBlockBreakExploits || world.getBlockState(headPos) == Blocks.PISTON_HEAD.defaultBlockState().setValue(FACING, enumdirection)) { // double check to make sure we're not a headless piston.
-                    world.removeBlock(headPos, false);
+                // Augustus start - Toggle Protect Bedrock and End Portal Frames from being destroyed
+                if (!world.augustusConfig.preventdestroyingunbreakableblocks) {
+                    world.removeBlock(pos.relative(enumdirection), false);
                 } else {
-                    ((ServerLevel)world).getChunkSource().blockChanged(headPos); // ... fix client desync
+                    // Paper start - fix headless pistons breaking blocks
+                    BlockPos headPos = pos.relative(enumdirection);
+                    if (io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPermanentBlockBreakExploits || world.getBlockState(headPos) == Blocks.PISTON_HEAD.defaultBlockState().setValue(FACING, enumdirection)) { // double check to make sure we're not a headless piston.
+                        world.removeBlock(headPos, false);
+                    } else {
+                        ((ServerLevel)world).getChunkSource().blockChanged(headPos); // ... fix client desync
+                    }
+                    // Paper end - fix headless pistons breaking blocks
                 }
-                // Paper end - fix headless pistons breaking blocks
+                // Augustus end - Toggle Protect Bedrock and End Portal Frames from being destroyed
             }
 
             world.playSound((Player) null, pos, SoundEvents.PISTON_CONTRACT, SoundSource.BLOCKS, 0.5F, world.random.nextFloat() * 0.15F + 0.6F);
diff --git a/src/main/java/net/minecraft/world/level/portal/PortalForcer.java b/src/main/java/net/minecraft/world/level/portal/PortalForcer.java
index 42212d4533ce25d1cfcf4c58f1fc88791d546cff..c9365d3c777307777ecb38900e992120deb49c1a 100644
--- a/src/main/java/net/minecraft/world/level/portal/PortalForcer.java
+++ b/src/main/java/net/minecraft/world/level/portal/PortalForcer.java
@@ -242,13 +242,17 @@ public class PortalForcer {
         for (int j = -1; j < 3; ++j) {
             for (int k = -1; k < 4; ++k) {
                 temp.setWithOffset(pos, portalDirection.getStepX() * j + enumdirection1.getStepX() * distanceOrthogonalToPortal, k, portalDirection.getStepZ() * j + enumdirection1.getStepZ() * distanceOrthogonalToPortal);
-                // Paper start - prevent destroying unbreakable blocks
-                if (!io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPermanentBlockBreakExploits) {
-                    if (!this.level.getBlockState(temp).isDestroyable()) {
-                        return false;
+                // Augustus start - Toggle Protect Bedrock and End Portal Frames from being destroyed
+                if (this.level.augustusConfig.preventdestroyingunbreakableblocks) {
+                    // Paper start - prevent destroying unbreakable blocks
+                    if (!io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPermanentBlockBreakExploits) {
+                        if (!this.level.getBlockState(temp).isDestroyable()) {
+                            return false;
+                        }
                     }
+                    // Paper end - prevent destroying unbreakable blocks
                 }
-                // Paper end - prevent destroying unbreakable blocks
+                // Augustus end - Toggle Protect Bedrock and End Portal Frames from being destroyed
                 if (k < 0 && !this.level.getBlockState(temp).isSolid()) {
                     return false;
                 }
diff --git a/src/main/java/org/skydom/server/chosen/AugustusWorldConfig.java b/src/main/java/org/skydom/server/chosen/AugustusWorldConfig.java
index 33b1d1ac84ef74abf9f3acde8e4f1a6b8c690c26..378458897c24fd4863b74e4baadabff15346c2bc 100644
--- a/src/main/java/org/skydom/server/chosen/AugustusWorldConfig.java
+++ b/src/main/java/org/skydom/server/chosen/AugustusWorldConfig.java
@@ -124,10 +124,12 @@ public class AugustusWorldConfig {
     public boolean safeteleporting = true;
     public boolean sandduping = false;
     public boolean tripwireupdate = false;
+    public boolean preventdestroyingunbreakableblocks = true;
 
     private void Settings() {
         safeteleporting = getBoolean("settings.safe-teleportation", safeteleporting);
         sandduping = getBoolean("settings.sand-duping", sandduping);
         tripwireupdate = getBoolean("settings.tripwire-update", tripwireupdate);
+        preventdestroyingunbreakableblocks = getBoolean("settings.prevent-destroying-unbreakable-blocks", preventdestroyingunbreakableblocks);
     }
 }
